libpseudodev/pseudodev.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Phoenix-RTOS
+ *
+ * pseudo devices
+ *
+ * Copyright 2023 Phoenix Systems
+ * Author: Gerard Swiderski
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/msg.h>
+#include <sys/file.h>
+#include <sys/time.h>
+#include "pseudodev.h"
+
+
+static struct {
+	unsigned int randSeed;
+} pseudo_common;
+
+
+static int pseudo_open(int id, int flags)
+{
+	(void)flags;
+
+	switch (id) {
+		case pseudo_idNull:
+		case pseudo_idZero:
+		case pseudo_idFull:
+		case pseudo_idRandom:
+			return EOK;
+
+		default:
+			break;
+	}
+
+	return -ENODEV;
+}
+
+
+static int pseudo_close(int id)
+{
+	switch (id) {
+		case pseudo_idNull:
+		case pseudo_idZero:
+		case pseudo_idFull:
+		case pseudo_idRandom:
+			return EOK;
+
+		default:
+			break;
+	}
+
+	return -ENODEV;
+}
+
+
+static ssize_t pseudo_read(int id, void *buf, size_t count)
+{
+	switch (id) {
+		case pseudo_idNull:
+			return (ssize_t)0;
+
+		case pseudo_idZero:
+		case pseudo_idFull:
+			memset(buf, 0, count);
+			return (ssize_t)count;
+
+		case pseudo_idRandom:
+			for (size_t ofs = 0; ofs < count; ofs += sizeof(pseudo_common.randSeed)) {
+				rand_r(&pseudo_common.randSeed);
+				memcpy((unsigned char *)buf + ofs, &pseudo_common.randSeed,
+					((count - ofs) > sizeof(pseudo_common.randSeed)) ? sizeof(pseudo_common.randSeed) : count - ofs);
+			}
+			return (ssize_t)count;
+
+		default:
+			break;
+	}
+
+	return -ENODEV;
+}
+
+
+static ssize_t pseudo_write(int id, const void *buf, size_t count)
+{
+	(void)buf;
+
+	switch (id) {
+		case pseudo_idNull:
+		case pseudo_idZero:
+		case pseudo_idRandom:
+			return (ssize_t)count;
+
+		case pseudo_idFull:
+			return -ENOSPC;
+
+		default:
+			break;
+	}
+
+	return -ENODEV;
+}
+
+
+int pseudo_handleMsg(msg_t *msg, int id)
+{
+	switch (msg->type) {
+		case mtOpen:
+			msg->o.io.err = pseudo_open(id, msg->i.openclose.flags);
+			return 0;
+
+		case mtClose:
+			msg->o.io.err = pseudo_close(id);
+			return 0;
+
+		case mtRead:
+			msg->o.io.err = pseudo_read(id, msg->o.data, msg->o.size);
+			return 0;
+
+		case mtWrite:
+			msg->o.io.err = pseudo_write(id, msg->i.data, msg->i.size);
+			return 0;
+
+		default:
+			break;
+	}
+
+	return -ENOSYS;
```

Comment:
Why not `msg->o.io.err = -ENOSYS;`? API is inconsistent and forces outside correction.
We can handle mtGet/SetAttr and mtDevCtl separately to be super correct

Reply:
The path `msg->o.io.err` is not always correct, e.g. `mtGetAttr` and `mtSetAttr` are no longer handled (I removed after [comment](https://github.com/phoenix-rtos/phoenix-rtos-devices/pull/428#discussion_r1346993503)) so the `default` switch-case clasuse would be incorrect (should be `msg->o.attr.err` instead of `msg->o.io.err`), the function `pseudo_handleMsg()` returns error if something has not been handled and requires further handling by the calling code (multidrv implementation). I removed all not handled/unsupported `mt*`s to simplyfy API and delegated handling to the driver (because it knows what to do in the default case).

In multidrv:
```C
static void multi_dispatchMsg(msg_t *msg)
...
		case id_pseudoRandom:
			if (pseudo_handleMsg(msg, multi2pseudo(id)) < 0) {
				multi_handleError(msg, -EPERM);
			}
			break;
...
```